x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharFormatter.java

```diff
@@ -0,0 +1,376 @@
+            return "nd";
+        } else if (mod10 == 3 && mod100 != 13) {
+            return "rd";
+        } else {
+            return "th";
+        }
+    }
+
+    private static String appendOrdinalSuffix(String defaultSuffix, String s) {
+        try {
+            // the Y,YYY pattern might can cause problems with the parsing, but thankfully the last 3
+            // characters is enough to calculate the suffix
+            int i = Integer.parseInt(lastNCharacter(s, 3));
+            final boolean upperCase = defaultSuffix.equals(defaultSuffix.toUpperCase(Locale.ROOT));
+            return s + (upperCase ? ordinalSuffix(i).toUpperCase(Locale.ROOT) : ordinalSuffix(i));
+        } catch (NumberFormatException ex) {
+            return s + defaultSuffix;
+        }
+    }
+
+    private static String removeLeadingZeros(String number) {
+        return String.valueOf(Integer.parseInt(number));
+    }
+
+    private static String removePadding(String paddedText) {
+        return paddedText.replaceAll(" +$", "");
+    }
+
+    private static String formatOffset(String offset) {
+        if (offset.equals("Z")) {
+            return "+00";
+        }
+        if (offset.matches("^[+-][0-9][0-9]00$")) {
+            offset = offset.substring(0, offset.length() - 2);
+        } else if (offset.matches("^[+-][0-9]{3,4}$")) {
+            offset = offset.substring(0, offset.length() - 2) + ":" + offset.substring(offset.length() - 2);
+        } else if (offset.matches("^[+-][0-9][0-9]:00$")) {
+            offset = offset.substring(0, offset.length() - 3);
+        }
+        return offset.substring(0, Math.min(offset.length(), 6));
+    }
+
+    private static String removeLeadingZerosFromOffset(String offset) {
+        if (offset.matches("[+-]0{1,2}")) {
+            return offset.substring(0, 2);
+        } else {
+            if (offset.startsWith("+0")) {
+                return "+" + offset.substring(2);
+            } else if (offset.startsWith("-0")) {
+                return "-" + offset.substring(2);
+            } else {
+                return offset;
+            }
+        }
+    }
+
+    private static String absoluteWeekBasedYear(TemporalAccessor t) {
+        int year = t.get(IsoFields.WEEK_BASED_YEAR);
+        year = year > 0 ? year : -(year - 1);
+        return String.format(Locale.ROOT, "%04d", year);
+    }
+
+    private static String firstDigitsOfNanos(String nano, int digits) {
+        return String.format(Locale.ROOT, "%09d", Integer.parseInt(nano)).substring(0, digits);
+    }
+
+    private static String lastNCharacter(String s, int n) {
+        return s.substring(Math.max(0, s.length() - n));
+    }
+
+    private static String zoneAbbreviationOf(TemporalAccessor temporalAccessor) {
+        String zone = ZoneId.from(temporalAccessor).getDisplayName(TextStyle.SHORT, Locale.ROOT);
+        return "Z".equals(zone) ? "UTC" : zone;
+    }
+
+    private static class Builder {
```

Comment:
Based on the code, it looks like the builder always ends up with text(), noFillMode or the default fill mode. If this assertion is correct, this behavior should be incapsulated in the builder to avoid calling build() and make these methods the final ones. Additionally aligned the names to better express what's going on:
text() -> noPadding()/removePadding()
text(fillMode) -> methodWithNameOfTheGivenFunction

Reply:
Refactored the builder `numeric()`, `numericWithLeadingZeros()`, `text()` and `offset()` replaced the `build()` method calls.